Rel_notes: storage (#56097)
openshift/openshift-docs: release_notes/ocp-4-13-release-notes.adoc

[id="ocp-4-13-storage"]
=== Storage

==== CSI volume health (Technology Preview)

> Container Storage Interface (CSI) volume health is currently enabled, and supported as a Technology Preview feature. However, {product-title} currently does not ship any CSI drivers that support this feature.

Reviewer: Which Jira epic does it map to ?

Reviewer 2: Each heading needs an [id="ocp-4-13-storage-blah"] thingy.

Reviewer 2: `does not ship any CSI drivers that support this feature.`
 
 What does this mean to the user?

Author: > Which Jira epic does it map to ?

https://issues.redhat.com/browse/OSDOCS-3793

Author: Done.

[id="ocp-4-13-storage"]
=== Storage

==== CSI volume health (Technology Preview)
Container Storage Interface (CSI) volume health is currently enabled, and supported as a Technology Preview feature. However, {product-title} currently does not ship any CSI drivers that support this feature.

==== CSI driver volume detach with non-graceful node shutdown (Technology Preview)

> Container Storage Interface (CSI) drivers can now automatically detach volumes when a node goes down non-gracefully. When a non-graceful node shutdown occurs, you can manually add an out-of-service taint on the node to allow volumes to automatically detach from the node. This feature is supported with Technology Preview status.

Reviewer 2: `you can manually add an out-of-service taint`

Does this needs to be done before the shutdown? I assume so; but it isn't 100% clear to me.

Author: Actually not. I'll correct the wording.

For more information, see ???.

==== VMware vSphere encryption support is generally available

> You can encrypt virtual machines (VMs) and persistent volumes (PVs) on {product-title} running on vSphere. This feature allows user-provisioned persistent storage to be encrypted over wire.

Reviewer 2: I assume folks that need to know will know what `over wire` means.

For more information, see ???.

==== VMware vSphere CSI topology support for multiple datacenters is generally available

> {product-title} 4.12 introduced the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple compute clusters, thus helping to avoid a single point of failure. {product-title} 4.13 introduces support for deploying over multiple datacenters as well.

Reviewer 2: s/as well/also

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-vsphere.adoc#persistent-storage-csi-vsphere-top-aware_persistent-storage-csi-vsphere[vSphere CSI topology].

==== Creating more than one default storage class is generally available

> {product-title} 4.13 allows you create more than one default storage class. This feature makes it easier to change the default storage class because you can create a second storage classes defined as the default, thus temporarily having two default storage classes, before removing default status from the previous default storage class.

Reviewer 2: s/second storage classes/second storage class

Consider splitting this second sentence into two. 

> This feature makes it easier to change the default storage class because you can create a second storage classes defined as the default. You then temporarily have two default storage classes before removing default status from the previous default storage class.

> While it is acceptable to have multiple default storage classes for a short time, you should ensure that eventually only one default storage class exists.

Reviewer 2: `eventually` seems too open-ended to me. 

> you should ensure you remove one of the default storage classes.

Author: Not sure we can give better guidance than that. There's not a time period that really makes sense to offer as guidance.

For more information, see ???.

==== Managing the default storage class is generally available 
{product-title} 4.13 introduces the `spec.storageClassState` field in the ClusterCSIDriver object, which allows you to manage the default storage class to accomplish several different objectives:

* Enforcing static provisioning by disabling dynamic provisioning.

* When you have other preferred storage classes, preventing the storage operator from re-creating the initial default storage class.

* Renaming, or otherwise changing, the default storage class

For more information, see ???.

==== Retroactive default StorageClass assignment (Technology Preview)

> There are some cases where the default storage classes are not assigned, or more than one default is defined. This feature allows you to retroactively assign a default storage class to correct theses situations. This feature is supported with Technology Preview status.

Reviewer 2: s/theses/these 

(I hate those meeces to pieces!)

Author: Hee hee.

For more information, see ???.

==== CSI inline ephemeral volumes is generally available

> Container Storage Interface (CSI) inline ephemeral volumes were introduced in {product-title} 4.5 as a Technology Preview feature, and allows you to define a Pod spec that creates inline ephemeral volumes when a pod is deployed and delete them when a pod is destroyed. This feature is now generally available.

Reviewer 2: s/ and allows/which allows

This feature is only available with supported Container Storage Interface (CSI) drivers.

For more information, see xref:../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc[CSI inline ephemeral volumes].

==== Automatic CSI migration for Microsoft Azure File auto migration is generally available

> Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for Azure File was provided in this feature in {product-title} 4.10, and {product-title} 4.13 now supports automatic migration for Azure File as generally available. CSI migration for Azure File is now enabled by default and requires no action by an administrator.

Reviewer 2: Maybe split the 4.10 and 4.13 sentences into two, so that `{product-title} 4.13 now supports automatic migration for Azure File as generally available.` stands alone, as this is the main point.

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

> While storage class referencing to the in-tree storage plugin will continue working, it is recommended that you switch the default storage class to the CSI storage class.

Reviewer 2: ISG says only use `while`to _refer to a period of time_.  Suggested `although` or `though`

`storage class referencing to the in-tree storage plugin` 
This is probably right. My ignorance of the subject makes it sound odd. 

s/will continue/continues

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

==== Automatic CSI migration for VMware vSphere auto migration is generally available

> Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for vSphere was provided in this feature in {product-title} 4.10, and {product-title} 4.13 now supports automatic migration for vSphere as generally available. CSI migration for vSphere is now enabled by default and requires no action by an administrator.

Reviewer 2: Same as above for splitting 4.10 and 4.13 phrases.

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

> While storage class referencing to the in-tree storage plugin will continue working, it is recommended that you switch the default storage class to the CSI storage class.

Reviewer 2: Same comment for `while`and `will continue`.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

==== Cross account support for AWS EFS CSI driver is generally available

> Cross account support allows you to have an {product-title} cluster in one Amazon Web Services (AWS) account and mount your file system in another AWS account using the AWS Elastic File System (EFS) Container Storage Interface (CSI) driver.

Reviewer 2: s/Cross account support/Cross-account support ??

Author: Really not sure about this, so I'm leaving it be.

[id="ocp-4-13-storage"]
=== Storage

[id="ocp-4-13-storage-user-managed-encryption"]
==== User-managed encryption is generally available
The user-managed encryption feature encrypts {product-title} node root volumes and the default storage class.

When starting up managed {product-title} clusters, you can pass a custom Key Management Service (KMS) key to the installation program to encrypt the root volumes for the nodes in the cluster. The installation program also encrypts the default storage class with this key. This allows you to encrypt storage volumes with your selected key, instead of the platform’s default account key.

This features supports the following storage types:

* Amazon Web Services (AWS) Elastic Block storage (EBS) (for more information, see ???)

* Microsoft Azure Disk storage (for more information, see ???)

* Google Cloud Platform (GCP) persistent disk (PD) storage (for more information, see ???)

[id="ocp-4-13-storage-csi-vol-health"]
==== CSI volume health (Technology Preview)

> Container Storage Interface (CSI) volume health is currently enabled, and supported as a Technology Preview feature. However, {product-title} currently does not ship any CSI drivers that support this feature.

Reviewer 3: Volume health tech preview has slipped to 4.14 or even later upstream.

Author: Removed.

[id="ocp-4-13-storage-csi-vol-detach-non-graceful-shutdown"]
==== CSI driver volume detach with non-graceful node shutdown (Technology Preview)

> Container Storage Interface (CSI) drivers can now automatically detach volumes when a node goes down non-gracefully. When a non-graceful node shutdown occurs, you can then manually add an out-of-service taint on the node to allow volumes to automatically detach from the node. This feature is supported with Technology Preview status.

Reviewer 4: LGTM

Author: Thanks!

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-vsphere.adoc#persistent-storage-csi-vsphere-top-aware_persistent-storage-csi-vsphere[vSphere CSI topology].

[id="ocp-4-13-storage-multiple-default-sc"]
==== Creating more than one default storage class is generally available
{product-title} 4.13 allows you create more than one default storage class. This feature makes it easier to change the default storage class because you can create a second storage class defined as the default. You then temporarily have two default storage classes before removing default status from the previous default storage class. While it is acceptable to have multiple default storage classes for a short time, you should ensure that eventually only one default storage class exists.

For more information, see ???.

[id="ocp-4-13-storage-managing-default-sc"]
==== Managing the default storage class is generally available 

> {product-title} 4.13 introduces the `spec.storageClassState` field in the `ClusterCSIDriver` object, which allows you to manage the default storage class to accomplish several different objectives:

Reviewer: the default storage class generated by openshift?

Author: Yes.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-vsphere.adoc#persistent-storage-csi-vsphere-top-aware_persistent-storage-csi-vsphere[vSphere CSI topology].

[id="ocp-4-13-storage-multiple-default-sc"]
==== Creating more than one default storage class is generally available
{product-title} 4.13 allows you create more than one default storage class. This feature makes it easier to change the default storage class because you can create a second storage class defined as the default. You then temporarily have two default storage classes before removing default status from the previous default storage class. While it is acceptable to have multiple default storage classes for a short time, you should ensure that eventually only one default storage class exists.

For more information, see ???.

[id="ocp-4-13-storage-managing-default-sc"]
==== Managing the default storage class is generally available 
{product-title} 4.13 introduces the `spec.storageClassState` field in the `ClusterCSIDriver` object, which allows you to manage the default storage class to accomplish several different objectives:

> * Enforcing static provisioning by disabling dynamic provisioning.

Reviewer 5: I feel this is not a good one to introduce this feature, wdyt? @ropatil010 @jsafrane

Reviewer: Limit storage consumption to static provisioning?

Author: @gcharot @duanwei33 Do you want to change the wording of the first bullet point to "Limit storage consumption to static provisioning" and move it to be the last bullet point?

Reviewer 5: Yes, I agree move it to the last. Thanks!

Author: Done.

* When you have other preferred storage classes, preventing the storage operator from re-creating the initial default storage class.

* Renaming, or otherwise changing, the default storage class

For more information, see ???.

[id="ocp-4-13-storage-retroactive-default-sc-assignment"]
==== Retroactive default StorageClass assignment (Technology Preview)
There are some cases where the default storage classes are not assigned, or more than one default is defined. This feature allows you to retroactively assign a default storage class to correct these situations. This feature is supported with Technology Preview status.

> 

Reviewer 6: https://issues.redhat.com/browse/STOR-1061
This story works when no default storage class exists.

Reviewer 7: @lpettyjo: As @chao007 pointed out: this feature is for cases when no default was set, more specifically if it was not set when PVC got created. This results in a PVC having no class which will get stuck in Pending phase. And with this feature enabled those PVC will get updated with default storage class **when/if** a default gets set later. Just a side node here: older versions than 4.13 did leave the PVC without storage class also if there was more than one default set, now one of them will be picked and assigned to PVC upon creation, the pick appears to be somewhat random to the user (it is age based, I would not dive into that here). But I think we could be more talkative here and explain both behavior changes briefly. I would write something like:

_There are some cases where it is desired to have more default storage classes configured or have no default set, for example when you need to transition to a different default storage class in your cluster. Creating any PVC (with no explicit storage class) while one of these configurations are in effect will cause that those PVCs will have no storage class assigned and dynamic provisioning will not take place without further manual intervention. OpenShift 4.13 improves this behavior so that having multiple default storage classes set will cause one class to be picked during PVC creation. This Technology Preview feature focuses on the second case and allows you to safely have no default set letting new PVCs be created without any assigned default. Those PVCs will remaining in Pending state and later, if you set a default storage class, those PVCs will get updated automatically with the current default and dynamic provisioning will take place._

Reviewer: The feature description is indeed confusing. It does not cover the case where "more than one default is defined".
Also it doesn't really "allow you" since it is automatic.

For more information, see ???.

[id="ocp-4-13-storage-csi-inline-ephemeral-vols-GA"]
==== CSI inline ephemeral volumes is generally available
Container Storage Interface (CSI) inline ephemeral volumes were introduced in {product-title} 4.5 as a Technology Preview feature, which allows you to define a pod spec that creates inline ephemeral volumes when a pod is deployed and delete them when a pod is destroyed. This feature is now generally available.

> This feature is only available with supported Container Storage Interface (CSI) drivers.

Reviewer 8: We'd better also metion that only when csi driver is not support `Generic ephemeral volumes`, it'll support CSI inline ephemeral and we recommend customers to use the `Generic ephemeral volumes`. WDYT?

Reviewer: AFAIK all drivers that supports dynamic provisioning supports generic ephemeral volume.

Reviewer 8: Yeah, all the playload dynamic provisioning csi driver support generic ephemeral volume, but looks like only few csi driver support the CSI inline ephemeral volumes(E.g. sharedresource csi driver and hostpath csi driver ), so maybe we could make the CSI inline ephemeral volumes support matrix clearly.

Reviewer: right, I haven't reviewed the CSI inline doc yet, if we want some kind of matrix or listing that would be there not in the release notes.

Reviewer: right, I haven't reviewed the CSI inline doc yet, if we want some kind of matrix or listing that would be there not in the release notes.

Reviewer 8: Got it. Thanks for clarifying!

Author: @Phaow @gcharot do we want this covered in the main docs? Agreed it doesn't belong in RNs.

Reviewer 9: IMO, this could go into https://github.com/openshift/openshift-docs/pull/47603, there are only 2 drivers that ship with OCP and support inline volumes.

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Although storage class referencing to the in-tree storage plugin will continue working, it is recommended that you switch the default storage class to the CSI storage class.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

[id="ocp-4-13-storage-aws-efs-cross-account-support"]
==== Cross account support for AWS EFS CSI driver is generally available
Cross account support allows you to have an {product-title} cluster in one Amazon Web Services (AWS) account and mount your file system in another AWS account using the AWS Elastic File System (EFS) Container Storage Interface (CSI) driver.

For more information, see ???.

[id="ocp-4-13-storage-fsgroup-to-csi-driver"]

> ==== Delegate FSGroup to CSI Driver instead of Kubelet (Technology Preview)

Reviewer 6: https://issues.redhat.com/browse/STOR-843 I think it is a GA feature.
@fabiand Can you help to double confirm?

Reviewer: Yes this is GA

[id="ocp-4-13-storage-lvms-dual-stack-support"]
==== Dual-stack support for {lvms-first}
In {product-title} {product-version}, {lvms} is supported in dual-stack for IPv4 and IPv6 network environments. For more information, see xref:../networking/ovn_kubernetes_network_provider/converting-to-dual-stack.adoc#nw-dual-stack-convert_converting-to-dual-stack[Converting to a dual-stack cluster network].

[id="ocp-4-13-storage-user-managed-encryption"]
==== User-managed encryption is generally available

> The user-managed encryption feature encrypts {product-title} node root volumes and the default storage class.

Reviewer 8: ```suggestion
The user-managed encryption feature encrypts {product-title} node root volumes and all the managed storage classes.
```
We also need update as we discussed in https://github.com/openshift/openshift-docs/pull/55014/files#r1152745609 .

Author: Done.

[id="ocp-4-13-storage-csi-inline-ephemeral-vols-GA"]
==== CSI inline ephemeral volumes is generally available
Container Storage Interface (CSI) inline ephemeral volumes were introduced in {product-title} 4.5 as a Technology Preview feature, which allows you to define a pod spec that creates inline ephemeral volumes when a pod is deployed and delete them when a pod is destroyed. This feature is now generally available.

This feature is only available with supported Container Storage Interface (CSI) drivers.

This feature also includes the CSI Volume Admission plug-in, which provides a mechanism where the use of an individual CSI driver capable of provisioning CSI ephemeral volumes can be restricted on pod admission. Administrators or distributions can add a `csi-ephemeral-volume-profile` label to a `CSIDriver` object, and the label is then inspected by the Admission plug-in and used in enforcement, warning, and audit decisions.

For more information, see xref:../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc[CSI inline ephemeral volumes].

[id="ocp-4-13-storage-csi-migration-azure-file-GA"]

> ==== Automatic CSI migration for Microsoft Azure File auto migration is generally available

Reviewer 9: ```suggestion
==== Automatic CSI migration for Microsoft Azure File is generally available
```
`auto migration` is a bit redundant IMO

Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for Azure File was provided in this feature in {product-title} 4.10. {product-title} 4.13 now supports automatic migration for Azure File as generally available. CSI migration for Azure File is now enabled by default and requires no action by an administrator.

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Although storage class referencing to the in-tree storage plug-in will continue working, it is recommended that you switch the default storage class to the CSI storage class.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

[id="ocp-4-13-storage-csi-migration-vsphere-GA"]

> ==== Automatic CSI migration for VMware vSphere auto migration is generally available

Reviewer 9: ```suggestion
==== Automatic CSI migration for VMware vSphere is generally available
```

Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for vSphere was provided in this feature in {product-title} 4.10. {product-title} 4.13 now supports automatic migration for vSphere as generally available. CSI migration for vSphere is now enabled by default and requires no action by an administrator.

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Although storage class referencing to the in-tree storage plug-in will continue working, it is recommended that you switch the default storage class to the CSI storage class.

> For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

Reviewer 9: Is it worth adding something here in the RN about the opt-in behavior for upgraded clusters in 4.13?

Author: Done.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-sc-manage.adoc[Managing the default storage class].

[id="ocp-4-13-storage-managing-default-sc"]
==== Managing the default storage class is generally available 
{product-title} 4.13 introduces the `spec.storageClassState` field in the `ClusterCSIDriver` object, which allows you to manage the default storage class generated by {product-title} to accomplish several different objectives:

* When you have other preferred storage classes, preventing the storage operator from re-creating the initial default storage class.

* Renaming, or otherwise changing, the default storage class

* Enforcing static provisioning by disabling dynamic provisioning.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-sc-manage.adoc[Managing the default storage class].

[id="ocp-4-13-storage-retroactive-default-sc-assignment"]
==== Retroactive default StorageClass assignment (Technology Preview)

> Previously, if there was no default storage class, persistent volumes claims (PVCs) that were created that requested the default storage class remained stranded in the pending state indefinitely. {product-title} can now retroactively assign a default storage class to these PVCs, so that they do not remain in the pending state. With this feature enabled, after a default storage class is created, or one of the existing storage classes is declared the default, these previously stranded PVCs are assigned the default storage class.

Reviewer 7: As for the PVCs being stranded indefinitely - this statement is not entirely true. Users had the option to delete and recreate the PVC, which would make it bind correctly. 

Otherwise this feature description looks good to me.

Author: @RomanBednar corrected.

[id="ocp-4-13-storage-lvms-in-disconnected-env"]
==== Support for {lvms} in disconnected environments

In {product-title} {product-version}, you can install {lvms} in disconnected environments. For more information, see xref:../storage/persistent_storage/persistent_storage_local/persistent-storage-using-lvms.adoc#lvms-installing-lvms-disconnected-env_logical-volume-manager-storage[Installing LVM Storage in a disconnected environment].

[id="ocp-4-13-storage-user-managed-encryption"]
==== User-managed encryption is generally available
The user-managed encryption feature allows you to provide keys during installation that encrypt {product-title} node root volumes, and enables all managed storage classes to use these keys to encrypt provisioned storage volumes. This allows you to encrypt storage volumes with your selected key, instead of the platform’s default account key.

This features supports the following storage types:

> * Amazon Web Services (AWS) Elastic Block storage (EBS) (for more information, see ???)

Reviewer: Missing link?

Author: Yes, the ??? is a placeholder that I use for links when i can't put them in because the content is not merged yet.

* Microsoft Azure Disk storage (for more information, see ???)

> * Google Cloud Platform (GCP) persistent disk (PD) storage (for more information, see ???)

Reviewer: Missing link?

[id="ocp-4-13-storage-vsphere-encryption-support"]
==== VMware vSphere encryption support is generally available
You can encrypt virtual machines (VMs) and persistent volumes (PVs) on {product-title} running on vSphere. 

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-vsphere.adoc#vsphere-pv-encryption[vSphere persistent disks encryption].

[id="ocp-4-13-storage-vpshere-top-support-multi-datacenters"]
==== VMware vSphere CSI topology support for multiple datacenters is generally available
{product-title} 4.12 introduced the ability to deploy {product-title} for vSphere on different zones and regions, which allows you to deploy over multiple compute clusters, thus helping to avoid a single point of failure. {product-title} 4.13 introduces support for deploying over multiple datacenters and to set up the topology using failure domains created during installation or post-installation.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-vsphere.adoc#persistent-storage-csi-vsphere-top-aware_persistent-storage-csi-vsphere[vSphere CSI topology].

[id="ocp-4-13-storage-multiple-default-sc"]

> ==== Creating more than one default storage class is generally available

Reviewer 9: @lpettyjo shouldn't this section reference the doc changes you made here, instead of `Managing the default storage class`?
https://github.com/openshift/openshift-docs/pull/55822/files

> ==== Creating more than one default storage class is generally available

Author: @dobsonj Yes and no. I was directing all of the SC stuff to that topic because it is actually a catch-all for all of the SC changes in 4.13. However, your point is valid in that it takes you to a top-level topic and this actual feature is somewhat buried in there, so I've changed it.

> {product-title} 4.13 allows you create more than one default storage class. This feature makes it easier to change the default storage class because you can create a second storage class defined as the default. You then temporarily have two default storage classes before removing default status from the previous default storage class. While it is acceptable to have multiple default storage classes for a short time, you should ensure that eventually only one default storage class exists.

Reviewer 7: Correct, having two default in the long run is still not recommended. Also we fire an alert after 10 minutes: https://github.com/openshift/cluster-storage-operator/blob/master/manifests/12_prometheusrules.yaml#L15-L27

Maybe we could just mention that users will get an alert if the don't ensure only one default is present?

* When you have other preferred storage classes, preventing the storage operator from re-creating the initial default storage class.

* Renaming, or otherwise changing, the default storage class

* Enforcing static provisioning by disabling dynamic provisioning.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-sc-manage.adoc[Managing the default storage class].

[id="ocp-4-13-storage-retroactive-default-sc-assignment"]
==== Retroactive default StorageClass assignment (Technology Preview)
Previously, if there was no default storage class, persistent volumes claims (PVCs) that were created that requested the default storage class remained stranded in the pending state indefinitely, unless you manually delete and recreate them. {product-title} can now retroactively assign a default storage class to these PVCs, so that they do not remain in the pending state. With this feature enabled, after a default storage class is created, or one of the existing storage classes is declared the default, these previously stranded PVCs are assigned to the default storage class.

This feature is supported with Technology Preview status.

> For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-sc-manage.adoc[Managing the default storage class].

Reviewer 7: @lpettyjo Similar to the [change](https://github.com/openshift/openshift-docs/pull/56097/files#r1189186657) above, I'd reference the [actual feature](https://github.com/openshift/openshift-docs/blob/6d34f08dd0d6ed73a12e8d6318b898d4610ed179/modules/persistent-storage-csi-sc-multiple-none.adoc?plain=1#L12) (titled "Absent default storage class") here too. That way users get to relevant feature details directly.

Otherwise SC feature notes lgtm.

Reviewer: +1 let's be consistent with the upstream name

Author: Done.

==== CSI inline ephemeral volumes is generally available
Container Storage Interface (CSI) inline ephemeral volumes were introduced in {product-title} 4.5 as a Technology Preview feature, which allows you to define a pod spec that creates inline ephemeral volumes when a pod is deployed and delete them when a pod is destroyed. This feature is now generally available.

This feature is only available with supported Container Storage Interface (CSI) drivers.

This feature also includes the CSI Volume Admission plug-in, which provides a mechanism where the use of an individual CSI driver capable of provisioning CSI ephemeral volumes can be restricted on pod admission. Administrators or distributions can add a `csi-ephemeral-volume-profile` label to a `CSIDriver` object, and the label is then inspected by the Admission plug-in and used in enforcement, warning, and audit decisions.

For more information, see xref:../storage/container_storage_interface/ephemeral-storage-csi-inline.adoc[CSI inline ephemeral volumes].

[id="ocp-4-13-storage-csi-migration-azure-file-GA"]
==== Automatic CSI migration for Microsoft Azure File is generally available

> Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for Azure File was provided in this feature in {product-title} 4.10. {product-title} 4.13 now supports automatic migration for Azure File as generally available. CSI migration for Azure File is now enabled by default and requires no action by an administrator.

Reviewer: WDYM by 

> Support for Azure File was provided in this feature in {product-title} 4.10

Azure file became GA in 4.11 see STOR-736. I think the sentence is confusing maybe we can make it more clear?

This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Although storage class referencing to the in-tree storage plug-in will continue working, it is recommended that you switch the default storage class to the CSI storage class.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

[id="ocp-4-13-storage-csi-migration-vsphere-GA"]
==== Automatic CSI migration for VMware vSphere is generally available
Starting with {product-title} 4.8, automatic migration for in-tree volume plugins to their equivalent Container Storage Interface (CSI) drivers became available as a Technology Preview feature. Support for vSphere was provided in this feature in {product-title} 4.10. {product-title} 4.13 now supports automatic migration for vSphere as generally available. CSI migration for vSphere is now enabled by default and requires no action by an administrator.

> This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Author: Done.

> This feature automatically translates in-tree objects to their counterpart CSI representations and should be completely transparent to users. Translated objects are not stored on disk, and user data is not migrated.

Reviewer 10: Yeah lets remove the last sentence.

For new installations of OpenShift Container Platform 4.13, or later, automatic migration is enabled by default. However, when upgrading from {product-title} 4.12, or earlier, to 4.13, automatic CSI migration for vSphere only occurs if you opt in. Carefully review the indicated consequences before opting in to migration. 

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-migration.adoc[CSI Automatic Migration].

[id="ocp-4-13-storage-aws-efs-cross-account-support"]
==== Cross account support for AWS EFS CSI driver is generally available
Cross account support allows you to have an {product-title} cluster in one Amazon Web Services (AWS) account and mount your file system in another AWS account using the AWS Elastic File System (EFS) Container Storage Interface (CSI) driver.

For more information, see xref:../storage/container_storage_interface/persistent-storage-csi-aws-efs.adoc#persistent-storage-csi-efs-cross-account_persistent-storage-csi-aws-efs[AWS EFS CSI cross account support].

[id="ocp-4-13-storage-fsgroup-to-csi-driver"]
==== Delegate FSGroup to CSI Driver instead of Kubelet is generally available

> This feature allows {product-title} to supply a pod's FSGroup to a Container Storage Interface (CSI) driver when a volume is mounted. However, currently, no {product-title} CSI drivers either support or take advantage of this feature.

Author: @jsafrane @gnufied Please weigh in. 

I was told that "We do have a CSI driver that takes advantage of this (azure-file-csi-driver), however, it's something hidden under the hood and users get their data on their volumes "as usual". It's yet another thing that's interesting only to CSI driver vendors."

> This feature allows {product-title} to supply a pod's FSGroup to a Container Storage Interface (CSI) driver when a volume is mounted. However, currently, no {product-title} CSI drivers either support or take advantage of this feature.

Reviewer 10: No I think @gcharot is correct. azurefile does uses this feature, so we should mention that azurefile *depends* on that feature.  We could simplify the wording of course and we don't have to mention all the inner working of CSI drivers, but saying that - "No CSI driver supports or take advantage of this feature" would be wrong.

> This feature allows {product-title} to supply a pod's FSGroup to a Container Storage Interface (CSI) driver when a volume is mounted. However, currently, no {product-title} CSI drivers either support or take advantage of this feature.

Author: Done.

(Azure file, VMware vSphere)
|Technology Preview
|Technology Preview
|General Availability

|CSI automatic migration

> (Azure Disk, OpenStack Cinder)

Reviewer 11: Similary For AWS EBS, GCP disk. General Availability starts from 4.12 version and currently it is marked from 4.13.

> (Azure Disk, OpenStack Cinder)

Reviewer: Correct 
4.11 : cinder & azure disk
4.12: GCE disk & EBS 
4.13: Azure file & vSphere
